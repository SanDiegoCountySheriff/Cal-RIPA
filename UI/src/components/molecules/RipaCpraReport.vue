<template>
  <v-container class="tw-mt-2" fluid>
    <v-layout row align-center nowrap class="tw-m-0">
      <v-flex xs12 md2>
        <div class="tw-ml-2">
          <ripa-date-picker
            label="From Date"
            class="tw-ml-2"
            @input="fromDateChange"
            :rules="dateRules"
          ></ripa-date-picker>
        </div>
      </v-flex>

      <v-flex xs12 md2>
        <div class="tw-ml-2">
          <ripa-date-picker
            label="To Date"
            class="tw-ml-2"
            @input="toDateChange"
            :rules="dateRules"
          ></ripa-date-picker>
        </div>
      </v-flex>

      <v-flex xs12 md2>
        <div class="tw-ml-2">
          <v-btn
            small
            color="primary"
            class="tw-ml-2"
            @click="createCpraReport(reportDates)"
            :disabled="!isValidDateRange"
            >Create</v-btn
          >
        </div>
      </v-flex>
    </v-layout>
    <v-card :loading="this.loading" class="cpra-report" flat>
      <v-card-title>CPRA Report</v-card-title>
      <v-card-text
        >Only stops accepted by the California DOJ are included on the
        report.</v-card-text
      >
      <v-card-text>
        <div v-for="(item, index) in cpraItems" :key="index">
          <ripa-list :item="item"></ripa-list>
        </div>
      </v-card-text>
      <div class="tw-ml-4 tw-mb-4">
        <v-btn
          @click="downloadReport(reportStats.fileName)"
          v-show="reportHasFilename"
          small
          color="primary"
          >Download Report</v-btn
        >
      </div>
    </v-card>
  </v-container>
</template>

<script>
import RipaDatePicker from '@/components/atoms/RipaDatePicker'
import RipaList from '@/components/molecules/RipaList'
import { dateNotInFuture } from '@/utilities/dates'

export default {
  name: 'ripa-cpra-report',

  components: { RipaDatePicker, RipaList },

  data() {
    return {
      fromDate: null,
      toDate: null,
      reportStats: {},
    }
  },

  computed: {
    dateRules() {
      return [
        v => !!v || 'A date is required',
        this.fromDate < this.toDate || 'From date must be before To Date',
        v => dateNotInFuture(v) || 'Date must not be in the future',
      ]
    },

    isValidDateRange() {
      return (
        this.fromDate !== null &&
        this.toDate !== null &&
        this.fromDate < this.toDate &&
        dateNotInFuture(this.fromDate) &&
        dateNotInFuture(this.toDate)
      )
    },

    reportDates() {
      return {
        fromDate: this.fromDate,
        toDate: this.toDate,
      }
    },

    cpraItems() {
      return this.reportStats?.cpraItems?.length > 0
        ? this.reportStats.cpraItems
        : []
    },

    reportHasFilename() {
      return this.reportStats?.fileName !== undefined
    },
  },

  methods: {
    init() {
      this.reportStats = this.items
    },

    fromDateChange(val) {
      this.fromDate = val
    },

    toDateChange(val) {
      this.toDate = val
    },

    createCpraReport(reportDates) {
      const reportParameters = {
        reportDates,
        officerName: `${this.user.firstName} ${this.user.lastName}`,
      }
      this.$emit('handleCreateCpraReport', reportParameters)
    },

    downloadReport(fileName) {
      this.$emit('handleDownloadCpraReport', fileName)
    },
  },

  created() {
    this.init()
  },

  watch: {
    items(val) {
      this.reportStats = val
    },
  },

  props: {
    items: {
      type: Object,
      default: () => {},
    },
    loading: {
      type: Boolean,
      default: false,
    },
    user: {
      type: Object,
      default: () => {},
    },
  },
}
</script>

<style lang="scss" scoped>
.cpra-report {
  width: 45%;
  border: 1px solid #ccc !important;
}
</style>
