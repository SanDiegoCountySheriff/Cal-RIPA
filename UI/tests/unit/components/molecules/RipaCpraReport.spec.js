import RipaCpraReport from '@/components/molecules/RipaCpraReport.vue'
import { shallowMount } from '@vue/test-utils'
import Vuetify from 'vuetify'

describe('Ripa CPRA Report', () => {
  let vuetify
  let wrapper

  beforeEach(() => {
    vuetify = new Vuetify()
  })

  afterEach(() => {
    wrapper.destroy()
  })

  const factory = propsData => {
    return shallowMount(RipaCpraReport, {
      vuetify,
      propsData: {
        ...propsData,
      },
    })
  }

  it('should match snapshot', () => {
    wrapper = factory()

    expect(wrapper.element).toMatchSnapshot()
  })
})
