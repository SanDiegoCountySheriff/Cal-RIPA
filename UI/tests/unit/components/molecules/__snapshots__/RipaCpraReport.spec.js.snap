// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Ripa CPRA Report should match snapshot 1`] = `
<div class="container container--fluid">
  <div class="v-card v-card--flat v-sheet theme--light">
    <div class="v-card__text">
      <div class="row">
        <div class="col-lg-3 col-6">
          <div class="v-input v-input--is-label-active theme--light v-text-field">
            <div class="v-input__control">
              <div class="v-input__slot">
                <div class="v-text-field__slot"><label for="input-4" class="v-label v-label--active theme--light" style="left: 0px; position: absolute;">From Date</label><input id="input-4" type="date"></div>
                <div class="v-input__append-inner">
                  <div></div>
                </div>
                <div class="v-input__append-inner">
                  <div class="v-input__icon v-input__icon--append"><i aria-hidden="true" class="v-icon notranslate mdi mdi-calendar theme--light"></i></div>
                </div>
              </div>
              <div class="v-text-field__details">
                <div class="v-messages theme--light">
                  <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div class="col-lg-3 col-6">
          <div class="v-input v-input--is-label-active theme--light v-text-field">
            <div class="v-input__control">
              <div class="v-input__slot">
                <div class="v-text-field__slot"><label for="input-9" class="v-label v-label--active theme--light" style="left: 0px; position: absolute;">To Date</label><input id="input-9" type="date"></div>
                <div class="v-input__append-inner">
                  <div></div>
                </div>
                <div class="v-input__append-inner">
                  <div class="v-input__icon v-input__icon--append"><i aria-hidden="true" class="v-icon notranslate mdi mdi-calendar theme--light"></i></div>
                </div>
              </div>
              <div class="v-text-field__details">
                <div class="v-messages theme--light">
                  <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div class="row">
        <div class="col-lg-3 col-6">
          <div class="ripa-switch" style="max-width: 250px;">
            <div class="v-input v-input--reverse v-input--expand v-input--hide-details theme--light v-input--selection-controls v-input--switch">
              <div class="v-input__control">
                <div class="v-input__slot">
                  <div class="v-input--selection-controls__input"><input aria-checked="false" id="input-14" role="switch" type="checkbox" aria-disabled="false" value="">
                    <div class="v-input--selection-controls__ripple"></div>
                    <div class="v-input--switch__track theme--light"></div>
                    <div class="v-input--switch__thumb theme--light">
                      <transition-stub name="fab-transition" mode="out-in"></transition-stub>
                    </div>
                  </div><label for="input-14" class="v-label theme--light" style="left: 0px; position: relative;">
                    Include Officer Information
                  </label>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div class="col-lg-3 col-6">
          <div class="ripa-switch" style="max-width: 250px;">
            <div class="v-input v-input--reverse v-input--expand v-input--hide-details theme--light v-input--selection-controls v-input--switch">
              <div class="v-input__control">
                <div class="v-input__slot">
                  <div class="v-input--selection-controls__input"><input aria-checked="false" id="input-17" role="switch" type="checkbox" aria-disabled="false" value="">
                    <div class="v-input--selection-controls__ripple"></div>
                    <div class="v-input--switch__track theme--light"></div>
                    <div class="v-input--switch__thumb theme--light">
                      <transition-stub name="fab-transition" mode="out-in"></transition-stub>
                    </div>
                  </div><label for="input-17" class="v-label theme--light" style="left: 0px; position: relative;">
                    Include Beat Information
                  </label>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div class="row">
        <div class="col"><button type="button" disabled="disabled" class="v-btn v-btn--disabled v-btn--has-bg theme--light v-size--default"><span class="v-btn__content">
            Create
          </span></button></div>
      </div>
      <div class="row">
        <div class="col-lg-6 col-12">
          <div class="v-card v-sheet v-sheet--outlined theme--light">
            <div class="v-card__title">CPRA Report</div>
            <div class="v-card__text">
              Only stops accepted by the California DOJ are included on the
              report.
            </div>
            <div class="v-card__text"></div>
            <div class="v-card__actions">
              <div class="spacer"></div> <button type="button" class="v-btn v-btn--text theme--light v-size--default primary--text" style="display: none;"><span class="v-btn__content">
                Download Report
              </span></button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
